File: src/main/java/org/cubectl/identity/auth/AuthController.java
package org.cubectl.identity.auth;

import java.util.Set;
import java.util.UUID;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/auth")
public class AuthController {

    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @PostMapping("/register")
    public ResponseEntity<AuthResponse> register(@Valid @RequestBody AuthRequest request) {
        AuthService.TokenResponse token = authService.register(request.username(), request.password());
        return ResponseEntity.status(HttpStatus.CREATED).body(AuthResponse.from(token));
    }

    @PostMapping("/login")
    public ResponseEntity<AuthResponse> login(@Valid @RequestBody AuthRequest request) {
        return ResponseEntity.ok(AuthResponse.from(authService.login(request.username(), request.password())));
    }

    @PostMapping("/logout")
    public ResponseEntity<Void> logout(@RequestBody(required = false) LogoutRequest request) {
        authService.logout(request == null ? null : request.refreshToken());
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/refresh")
    public ResponseEntity<AuthResponse> refresh(@Valid @RequestBody RefreshRequest request) {
        return ResponseEntity.ok(AuthResponse.from(authService.refresh(request.refreshToken())));
    }

    @GetMapping("/me")
    public ResponseEntity<MeResponse> me(@AuthenticationPrincipal AuthenticatedUser user) {
        return ResponseEntity.ok(new MeResponse(user.userId(), user.username(), user.permissions()));
    }

    public record AuthRequest(
            @NotBlank @Size(min = 3, max = 64) String username,
            @NotBlank @Size(min = 8, max = 128) String password
    ) {
    }

    public record RefreshRequest(@NotBlank String refreshToken) {
    }

    public record LogoutRequest(String refreshToken) {
    }

    public record AuthResponse(String accessToken, String refreshToken, String tokenType) {
        static AuthResponse from(AuthService.TokenResponse token) {
            return new AuthResponse(token.accessToken(), token.refreshToken(), token.tokenType());
        }
    }

    public record MeResponse(UUID userId, String username, Set<String> permissions) {
    }
}

File: src/main/kotlin/org/cubectl/identity/auth/AuthController.kt
package org.cubectl.identity.auth

import jakarta.validation.Valid
import org.cubectl.identity.auth.dto.AuthRequest
import org.cubectl.identity.auth.dto.AuthResponse
import org.cubectl.identity.auth.dto.LogoutRequest
import org.cubectl.identity.auth.dto.MeResponse
import org.cubectl.identity.auth.dto.RefreshRequest
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/auth")
class AuthController(
    private val authService: AuthService,
) {

    @PostMapping("/register")
    fun register(@Valid @RequestBody request: AuthRequest): ResponseEntity<AuthResponse> {
        val token = authService.register(request.username, request.password)
        return ResponseEntity.status(HttpStatus.CREATED).body(AuthResponse.from(token))
    }

    @PostMapping("/login")
    fun login(@Valid @RequestBody request: AuthRequest): ResponseEntity<AuthResponse> =
        ResponseEntity.ok(AuthResponse.from(authService.login(request.username, request.password)))

    @PostMapping("/logout")
    fun logout(@RequestBody(required = false) request: LogoutRequest?): ResponseEntity<Void> {
        authService.logout(request?.refreshToken)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("/refresh")
    fun refresh(@Valid @RequestBody request: RefreshRequest): ResponseEntity<AuthResponse> =
        ResponseEntity.ok(AuthResponse.from(authService.refresh(request.refreshToken)))

    @GetMapping("/me")
    fun me(@AuthenticationPrincipal user: AuthenticatedUser): ResponseEntity<MeResponse> =
        ResponseEntity.ok(MeResponse(user.userId, user.username, user.permissions))
}
